src/buf/fixed/buffers.rs

```diff
@@ -1,113 +1,12 @@
-        let orig_cap = iovecs.capacity();
-        mem::forget(iovecs);
-        FixedBuffers {
-            raw_bufs,
-            states,
-            orig_cap,
-        }
-    }
-
-    // If the indexed buffer is free, changes its state to checked out and
-    // returns its data. If the buffer is already checked out, returns None.
-    pub(crate) fn check_out(&mut self, index: usize) -> Option<(iovec, usize)> {
-        let iovecs_ptr = self.raw_bufs;
-        self.states.get_mut(index).and_then(|state| match *state {
-            BufState::Free { init_len } => {
-                *state = BufState::CheckedOut;
-                // Safety: the allocated array under the pointer is valid
-                // for the lifetime of self, the index is inside the array
-                // as checked by Vec::get_mut above, called on the array of
-                // states that has the same length.
-                let iovec = unsafe { iovecs_ptr.as_ptr().add(index).read() };
-                Some((iovec, init_len))
-            }
-            BufState::CheckedOut => None,
-        })
-    }
+// Abstracts management of fixed buffers in a buffer registry.
+pub(crate) trait FixedBuffers {
+    // Provides access to the raw buffers as a slice of iovec.
+    fn iovecs(&self) -> &[iovec];
 
     // Sets the indexed buffer's state to free and records the updated length
     // of its initialized part. The buffer addressed must be in the checked out
-    // state, otherwise this function may panic.
-    pub(crate) fn check_in(&mut self, index: usize, init_len: usize) {
-        let state = self.states.get_mut(index).expect("invalid buffer index");
-        debug_assert!(
-            matches!(state, BufState::CheckedOut),
-            "the buffer must be checked out"
-        );
-        *state = BufState::Free { init_len };
-    }
-
-    pub(crate) fn iovecs(&self) -> &[iovec] {
-        // Safety: the raw_bufs pointer is valid for the lifetime of self,
-        // the slice length is valid by construction.
-        unsafe { slice::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len()) }
-    }
-}
-
-impl Drop for FixedBuffers {
-    fn drop(&mut self) {
-        let iovecs = unsafe {
-            Vec::from_raw_parts(self.raw_bufs.as_ptr(), self.states.len(), self.orig_cap)
-        };
-        for (i, iovec) in iovecs.iter().enumerate() {
-            match self.states[i] {
-                BufState::Free { init_len } => {
-                    let ptr = iovec.iov_base as *mut u8;
-                    let cap = iovec.iov_len;
-                    let v = unsafe { Vec::from_raw_parts(ptr, init_len, cap) };
-                    mem::drop(v);
-                }
-                BufState::CheckedOut => unreachable!("all buffers must be checked in"),
-            }
-        }
-    }
+    // state, otherwise this function will panic.
+    fn check_in(&mut self, buf_index: u16, init_len: usize);
```

Comment:
I wonder if this method needs to be declared unsafe.

Surely the implementation does not need to do anything unsafe, but on the next check out, the buffer's initialized length will be set to the provided value.
So the caller of this method needs a speed bump like the one you get with `Vec::set_len`.

Reply:
That sounds good to me. I like speed bumps.